codersdk/workspacesdk/connector_internal_test.go

```diff
@@ -142,6 +146,217 @@ func TestTailnetAPIConnector_UplevelVersion(t *testing.T) {
+
+	var (
+		websocketConnCh   = make(chan *websocket.Conn, 64)
+		expectResumeToken = ""
+	)
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Accept a resume_token query parameter to use the same peer ID. This
+		// behavior matches the actual client coordinate route.
+		var (
+			peerID      = uuid.New()
+			resumeToken = r.URL.Query().Get("resume_token")
+		)
+		t.Logf("received resume token: %s", resumeToken)
+		assert.Equal(t, expectResumeToken, resumeToken)
+		if resumeToken != "" {
+			peerID, err = resumeTokenProvider.VerifyResumeToken(resumeToken)
+			assert.NoError(t, err, "failed to parse resume token")
+			if err != nil {
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: CoordinateAPIInvalidResumeToken,
+					Detail:  err.Error(),
+					Validations: []codersdk.ValidationError{
+						{Field: "resume_token", Detail: CoordinateAPIInvalidResumeToken},
+					},
+				})
+				return
+			}
+		}
+
+		sws, err := websocket.Accept(w, r, nil)
+		if !assert.NoError(t, err) {
+			return
+		}
+		testutil.RequireSendCtx(ctx, t, websocketConnCh, sws)
+		ctx, nc := codersdk.WebsocketNetConn(r.Context(), sws, websocket.MessageBinary)
+		err = svc.ServeConnV2(ctx, nc, tailnet.StreamID{
+			Name: "client",
+			ID:   peerID,
+			Auth: tailnet.ClientCoordinateeAuth{AgentID: agentID},
+		})
+		assert.NoError(t, err)
+	}))
+
+	fConn := newFakeTailnetConn()
+
+	newTickerTrap := clock.Trap().NewTicker("tailnetAPIConnector", "refreshToken")
+	tickerResetTrap := clock.Trap().TickerReset("tailnetAPIConnector", "refreshToken", "reset")
+	defer newTickerTrap.Close()
+	uut := newTailnetAPIConnector(ctx, logger, agentID, svr.URL, clock, &websocket.DialOptions{})
+	uut.runConnector(fConn)
+
+	// Fetch first token. We don't need to advance the clock since we use a
+	// channel with a single item to immediately fetch.
+	trappedTicker := newTickerTrap.MustWait(ctx)
+	trappedTicker.Release()
+	// We call ticker.Reset after each token fetch to apply the refresh duration
+	// requested by the server.
+	trappedReset := tickerResetTrap.MustWait(ctx)
+	trappedReset.Release()
+	require.NotNil(t, uut.resumeToken)
+	originalResumeToken := uut.resumeToken.Token
+
+	// Fetch second token.
+	waiter := clock.Advance(trappedReset.Duration)
+	waiter.MustWait(ctx)
+	trappedReset = tickerResetTrap.MustWait(ctx)
+	trappedReset.Release()
+	require.NotNil(t, uut.resumeToken)
+	require.NotEqual(t, originalResumeToken, uut.resumeToken.Token)
+	expectResumeToken = uut.resumeToken.Token
+	t.Logf("expecting resume token: %s", expectResumeToken)
+
+	// Sever the connection and expect it to reconnect with the resume token.
+	wsConn := testutil.RequireRecvCtx(ctx, t, websocketConnCh)
+	_ = wsConn.Close(websocket.StatusGoingAway, "test")
+
+	// Wait for the resume token to be refreshed.
+	trappedTicker = newTickerTrap.MustWait(ctx)
+	trappedTicker.Release()
+	waiter = clock.Advance(trappedTicker.Duration)
```

Comment:
I think this causes another instance of the race where we could refresh twice, but only trap it once, since there is another immediate refresh upon reconnection to the tailnet API.  So, you don't need to advance the clock to get it to refresh, and, the refresh might happen _before_ you advance the clock here, which will result in generating an identical token as before.

I think the solution is to advance the clock just a lil' bit while you are holding the `NewTicker` call trapped.  1 second is enough to cause a different JWT.  To the code under test, this is equivalent to some time passing between the last refresh and the reconnection.

It'll look like this:

```
	// Wait for the resume token to be refreshed.
	trappedTicker = newTickerTrap.MustWait(ctx)
	// Advance the clock so that the JWT will be different (different exp).
	// We do this before the call to NewTicker returns so that only the initial refresh
	// can trigger.
	clock.Advance(time.Second).MustWait(ctx)
	trappedTicker.Release()
	trappedReset = tickerResetTrap.MustWait(ctx)
	trappedReset.Release()
```